### PR TITLE
Change "filename" string to "files" list, and implement optional-dependencies

### LINKED ISSUE #6: Support for features / optional dependencies
Hey, great plugin!

For example this https://github.com/aws/sagemaker-python-sdk/blob/49c664893339b42815b2fe8a91d912194199e8ba/setup.py#L64-L72 might become:

```toml
[tool.hatch.metadata.hooks.requirements_txt.features]
local = ["requirements/extras/local_requirements.txt"]
scipy = ["requirements/extras/scipy_requirements.txt"]
all = [
  "requirements/extras/local_requirements.txt",
  "requirements/extras/scipy_requirements.txt",
]
test = [
  "requirements/extras/local_requirements.txt",
  "requirements/extras/scipy_requirements.txt",
  "requirements/extras/test_requirements.txt",
]
```

& maybe the top-level option should become an array too?

```toml
[tool.hatch.metadata.hooks.requirements_txt]
files = ["requirements.txt"]
```

### PR DESCRIPTION
**Pull request by maresb**
Closes #6

### CODE REVIEW

File: hatch_requirements_txt/__init__.py [hunk 1]
```diff
@@ -68,6 +68,33 @@ def parse_requirements(requirements: Iterable[str]) -> Tuple[List[Requirement],
...
+	:param files:
+
+	:return: The requirements, and a list of commented lines.
+	"""
+
+	all_parsed_requirements: List[Requirement] = []
+	all_comments = []
+
+	if not isinstance(files, List):
+		raise ValueError(f"Requirements files must be a list, but got {type(files)}: {files}.")
```
Comment (domdfcoding) [inline hatch_requirements_txt/__init__.py:84]: I feel like this (and the one a few lines down) should be `TypeError` instead, as the issue is that the *type* is wrong.

**Review by domdfcoding — COMMENTED** [review]
I'm happy with the main code changes, but I haven't had a chance to look at the tests yet.

File: hatch_requirements_txt/__init__.py (hunk 1 above)
Comment (maresb) [reply hatch_requirements_txt/__init__.py]: Yes, good call. Fixed.

**Comment by maresb** [comment]
@domdfcoding I didn't realize that the coverage tests were failing. (I don't think I got notified when repo-helper added the tags.)

I added a few more tests. Could you please restart the CI?

**Comment by maresb** [comment]
The flake8 failure looks like it may be caused by a discrepancy between the settings for `formate` and `flake8`.

File: tests/test_metadata.py [hunk 2]
```diff
@@ -33,51 +31,103 @@
...
 "Bug Tracker" = "https://github.com/pypa/sampleproject/issues"
-
-[tool.hatch.metadata.hooks.requirements_txt]
-filename = "requirements.txt"
 """
 
 
-def test_wheel(tmp_pathplus: PathPlus):
+def get_pkginfo(tmp_pathplus: PathPlus, build_func: Callable,
+				pyproject_toml: str) -> Union[pkginfo.SDist, pkginfo.Wheel]:
```
Comment (domdfcoding) [inline tests/test_metadata.py:38]:
```suggestion
def get_pkginfo(
		tmp_pathplus: PathPlus,
		build_func: Callable,
		pyproject_toml: str,
		) -> Union[pkginfo.SDist, pkginfo.Wheel]: 
```

That should make both formate and flake8 happy.

**Comment by maresb** [comment]
Cool, thanks! Hopefully now the CI passes.

**Review by ofek — APPROVED** [review]

File: tests/test_metadata.py [hunk 3]
```diff
@@ -33,51 +31,106 @@
...
+	assert info.requires_dist == [
+			"bar",
+			"baz>1",
+			"foo",
+			"colorama; platform_system == 'Windows' and extra == 'cli'",
+			"prompt-toolkit; extra == 'cli'",
+			"cryptography; extra == 'crypto'",
+			"pyjwt; extra == 'crypto'",
+			"orjson; extra == 'fastjson'"
+			]
```
Comment (domdfcoding) [inline tests/test_metadata.py:133]: Could you also add a test where the optional requirements files are in a subdirectory? I've seen some projects that have it like:

```
requirements/dev.txt
requirements/docs.txt
requirements/tests.txt
```

Looking at the code it should work but it's worth testing.

**Review by domdfcoding — CHANGES_REQUESTED** [review]

File: tests/test_metadata.py (hunk 3 above)
Comment (maresb) [reply tests/test_metadata.py:133]: Ya, I'll try to do that this evening.

Comment (maresb) [reply tests/test_metadata.py:149]: @domdfcoding, I have added the requested test.

**Review by ofek — APPROVED** [review]

**Comment by maresb** [comment]
Thanks @domdfcoding for the thorough review! I'm looking forward to the next release.

**Pull request merged**